lib/trino-collect/src/main/java/io/trino/collect/cache/EvictableCache.java

```diff
@@ -125,18 +124,22 @@ public V get(K key)
     {
         Token<K> newToken = new Token<>(key);
         Token<K> token = tokens.computeIfAbsent(key, ignored -> newToken);
-        try {
-            return dataCache.get(token);
-        }
-        catch (Throwable e) {
-            if (newToken == token) {
+        if (token == newToken) {
+            try {
+                return dataCache.get(newToken);
+            }
+            catch (Throwable e) {
                 // Failed to load and it was our new token persisted in tokens map.
                 // No cache entry exists for the token (unless concurrent load happened),
                 // so we need to remove it.
                 tokens.remove(key, newToken);
+                throw e;
             }
-            throw e;
         }
+        // Do not invoke the loader on the old token. The old token may have been put into the cache
+        // using e.g. get(key, Callable) and may get rejected by the cache loader function.
+        // Pass new token to the cache loader function instead.
+        return dataCache.get(token, () -> dataCache.get(newToken));
```

Comment:
If the loader is called `dataCache.get(newToken)` the cache will have two entries:
- token -> data
- newToken -> data

but you only want one associated with `newToken` right?